File: problems-extend-adapt/1401-1500/1449_biggest-number-on-an-exact-budget/solution.py
from typing import List


class Solution:
    def biggestOnBudget(self, cost: List[int], target: int) -> str:
        NEG = -1
        dp = [NEG] * (target + 1)
        dp[0] = 0
        for t in range(1, target + 1):
            for c in cost:
                if c <= t and dp[t - c] != NEG:
                    if dp[t - c] + 1 > dp[t]:
                        dp[t] = dp[t - c] + 1
        if dp[target] == NEG:
            return "0"
        result = []
        remaining = target
        while remaining > 0:
            for digit in range(9, 0, -1):
                c = cost[digit - 1]
                if c <= remaining and dp[remaining - c] == dp[remaining] - 1:
                    result.append(str(digit))
                    remaining -= c
                    break
        return "".join(result)

File: problems-extend-adapt/1401-1500/1449_biggest-number-on-an-exact-budget/solution.go
func biggestOnBudget(cost []int, target int) string {
	dp := make([]int, target+1)
	for t := 1; t <= target; t++ {
		dp[t] = -1
	}
	for t := 1; t <= target; t++ {
		for _, c := range cost {
			if c <= t && dp[t-c] != -1 && dp[t-c]+1 > dp[t] {
				dp[t] = dp[t-c] + 1
			}
		}
	}
	if dp[target] == -1 {
		return "0"
	}
	result := make([]byte, 0, dp[target])
	remaining := target
	for remaining > 0 {
		for digit := 9; digit >= 1; digit-- {
			c := cost[digit-1]
			if c <= remaining && dp[remaining-c] == dp[remaining]-1 {
				result = append(result, byte('0'+digit))
				remaining -= c
				break
			}
		}
	}
	return string(result)
}
